### PR TITLE
add event handler framework

### PR DESCRIPTION
**Pull request by sivanzcw**
implementation of proposal https://github.com/volcano-sh/volcano/pull/1886

### CODE REVIEW

**Comment by hwdef** [comment]
lgtm overall. Thank you!

**Comment by Thor-wl** [comment]
/hold 
Thanks for your contribution. Since it's in the period of code froze for v1.5, I will cancel hold after the release is launched.

**Comment by william-wang** [comment]
please resolve the confliction firstly.

**Comment by sivanzcw** [comment]
fixed. @william-wang

File: cmd/scheduler/app/options/options.go [hunk 1]
```diff
@@ -111,6 +127,11 @@ func (s *ServerOption) AddFlags(fs *pflag.FlagSet) {
 		"Enable tracking of available storage capacity that CSI drivers provide; it is false by default")
 	fs.BoolVar(&s.EnableHealthz, "enable-healthz", false, "Enable the health check; it is false by default")
 	fs.BoolVar(&s.EnableMetrics, "enable-metrics", false, "Enable the metrics function; it is false by default")
+
+	fs.StringVar(&s.EventHandler.Name, "event-handler", DefaultEventHandlerName,
+		"The name of event handler. Empty string for default event handler.")
+	fs.StringVar(&s.EventHandler.ConfigFile, "event-handler-config", s.EventHandler.ConfigFile,
+		"The path to the event handler configuration file. Empty string for no configuration file.")
```
[the remark this answers is not in the document]
Comment (sivanzcw) [reply cmd/scheduler/app/options/options.go:136]: - Specifying  `event-handler-config` alone will be ignored.
- Keep the `event-handler-config` empty, if the custom eventHandler does not have parameter configuration requirements.

File: cmd/scheduler/app/options/options.go [hunk 2]
```diff
@@ -68,11 +73,21 @@ type ServerOption struct {
...
 
 // ServerOpts server options.
 var ServerOpts *ServerOption
 
+// EventHandlerConfiguration contains basically elements about event handler.
+type EventHandlerConfiguration struct {
+	// Name is the handler for event processing.
+	Name string
+	// ConfigFile is the path to the event handler configuration file.
+	ConfigFile string
```
Comment (william-wang) [inline cmd/scheduler/app/options/options.go:88]: what's the content of the ConfigFile?

Comment (sivanzcw) [reply cmd/scheduler/app/options/options.go:88]: developer custom defined according to external eventhandler requirements

File: pkg/scheduler/cache/cache.go [hunk 3]
```diff
@@ -85,6 +85,14 @@ type SchedulerCache struct {
 	schedulerName      string
 	nodeSelectorLabels map[string]string
 
+	// CustomInformers is an interface list for custom informers.
+	// It is the entry point for custom object event processing, which can be
+	// accessed to the scheduling framework by implementing the SharedInformer interface.
+	// These shared informer will start and run at the beginning of the scheduler.
+	// The CustomInformers will be used along with the EventHandler framework.
+	// Initialize the CustomInformers, while initializing the EventHandler object.
+	CustomInformers []cache.SharedIndexInformer
```
Comment (william-wang) [inline pkg/scheduler/cache/cache.go:94]: How to add CustomerInformers and use it? is there a example or instruction for it in design doc or somewhere?

Comment (sivanzcw) [reply pkg/scheduler/cache/cache.go:94]: please see test case in plugins_test.go
image

File: pkg/scheduler/cache/fake_eventhandler.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2021 The Kubernetes Authors.
```
Comment (william-wang) [inline pkg/scheduler/cache/fake_eventhandler.go:2]: replace 'Kubernetes' to `Volcano`

File: pkg/scheduler/cache/event.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2021 The Kubernetes Authors.
```
Comment (william-wang) [inline pkg/scheduler/cache/event.go:2]: replace 'Kubernetes' to 'Volcano'.

File: pkg/scheduler/cache/event.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
+/*
```
Comment (william-wang) [inline pkg/scheduler/cache/event.go:1]: it better change the file name from ‘event.go’ to 'event_handlers_interface'

Comment (sivanzcw) [reply pkg/scheduler/cache/event.go:1]: golang does not recommend using underscores to name the file name, except for test files
[https://go.dev/doc/effective_go#mixed-caps](url)

File: pkg/scheduler/cache/fake_eventhandler.go (hunk 4 above)
Comment (sivanzcw) [reply pkg/scheduler/cache/fake_eventhandler.go:2]: There are a large number of files in the stock code file using `kubernetes` in License, even [missing](https://github.com/volcano-sh/volcano/blob/master/pkg/scheduler/actions/elect/elect.go). Whether to consider providing script to refresh and verification?

File: pkg/scheduler/cache/event.go (hunk 5 above)
Comment (sivanzcw) [reply pkg/scheduler/cache/event.go]: done

File: pkg/scheduler/cache/event.go (hunk 6 above)
Comment (william-wang) [reply pkg/scheduler/cache/event.go:1]: I know it's a not a recommended way with cunderscores in filenames. From my point of view, an explicit filename with multiple words get more benifit. It's common in k8s. we can shorten it to two words 'event_interface'.

Comment (sivanzcw) [reply pkg/scheduler/cache/event.go]: modify the filename:
event.go -> eventhandlerinterface.go

**Pull request closed without merging**